packages/back-end/src/queryRunners/QueryRunner.ts

```diff
@@ -258,21 +339,78 @@ export abstract class QueryRunner<
     }
   }
 
+  public async executeQuery<
+    Rows extends RowsType,
+    ProcessedRows extends ProcessedRowsType
+  >(
+    doc: QueryInterface,
+    run: (query: string) => Promise<QueryResponse<Rows>>,
+    process: (rows: Rows) => ProcessedRows
+  ): Promise<void> {
+    // Update heartbeat for the query once every 30 seconds
+    // This lets us detect orphaned queries where the thread died
+    const timer = setInterval(() => {
+      updateQuery(doc, { heartbeat: new Date() }).catch((e) => {
+        logger.error(e);
+      });
+    }, 30000);
+
+    // Run the query in the background
+    logger.debug(`Start executing query in background: ${doc.id}`);
+    await updateQuery(doc, {
+      startedAt: new Date(),
+      status: "running",
+    });
+
+    run(doc.query)
+      .then(async ({ rows, statistics }) => {
+        clearInterval(timer);
+        logger.debug("Query succeeded");
+        await updateQuery(doc, {
+          finishedAt: new Date(),
+          status: "succeeded",
+          rawResult: rows,
+          result: process(rows),
+          statistics: statistics,
+        });
+        this.onQueryFinish();
+      })
+      .catch(async (e) => {
+        clearInterval(timer);
+        logger.debug("Query failed: " + e.message);
+        updateQuery(doc, {
+          finishedAt: new Date(),
+          status: "failed",
+          error: e.message,
+        })
+          .then(() => {
+            this.onQueryFinish();
+          })
+          .catch((e) => logger.error(e));
+      });
+  }
+
   public async startQuery<
-    Rows extends Record<string, string | boolean | number>[],
-    // eslint-disable-next-line
-    ProcessedRows extends Record<string, any>
+    Rows extends RowsType,
+    ProcessedRows extends ProcessedRowsType
   >(
     name: string,
     query: string,
-    run: (
-      query: string
-    ) => Promise<{ statistics?: QueryStatistics; rows: Rows }>,
+    dependencies: string[],
+    run: (query: string) => Promise<QueryResponse<Rows>>,
     process: (rows: Rows) => ProcessedRows
   ): Promise<QueryPointer> {
```

Comment:
With this many arguments, we should refactor to not use positional arguments.  Make it accept a single object argument with named keys.